File: app/src/main/java/com/chrisprime/primestationonecontrol/views/EmptyRecyclerView.java
package com.chrisprime.primestationonecontrol.views;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.util.AttributeSet;
import android.view.View;
import android.widget.ProgressBar;

import rx.functions.Action0;

public class EmptyRecyclerView extends RecyclerView {
    private View mEmptyView;
    private ProgressBar mProgressBar;

    private boolean mLoading = false;

    private Action0 mEmptyAction, mNonEmptyAction;

    final private AdapterDataObserver observer = new AdapterDataObserver() {
        @Override
        public void onChanged() {
            checkIfEmpty();
        }

        @Override
        public void onItemRangeInserted(int positionStart, int itemCount) {
            checkIfEmpty();
        }

        @Override
        public void onItemRangeRemoved(int positionStart, int itemCount) {
            checkIfEmpty();
        }
    };

    public EmptyRecyclerView(Context context) {
        super(context);
    }

    public EmptyRecyclerView(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public EmptyRecyclerView(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
    }

    public void setLoading(boolean loading) {
        mLoading = loading;
        if (mProgressBar != null) {
            mProgressBar.setVisibility(loading ? VISIBLE : GONE);
        }
        if (loading) {
            mEmptyView.setVisibility(GONE);
        } else {
            checkIfEmpty();
        }
    }

    public void checkIfEmpty() {
        if (getAdapter() == null) {
            return;
        }
        final boolean emptyViewVisible = getAdapter().getItemCount() == 0;
        if (mEmptyView != null) {
            mEmptyView.setVisibility(emptyViewVisible && !mLoading ? VISIBLE : GONE);
            setVisibility(emptyViewVisible || mLoading ? GONE : VISIBLE);
        }

        if (emptyViewVisible && mEmptyAction != null) {
            mEmptyAction.call();
        } else if (!emptyViewVisible && mNonEmptyAction != null) {
            mNonEmptyAction.call();
        }
    }


    public void setEmptyAction(Action0 emptyAction) {
        mEmptyAction = emptyAction;
    }

    public void setNonEmptyAction(Action0 nonEmptyAction) {
        mNonEmptyAction = nonEmptyAction;
    }

    @Override
    public void setAdapter(Adapter adapter) {
        final Adapter oldAdapter = getAdapter();
        if (oldAdapter != null) {
            oldAdapter.unregisterAdapterDataObserver(observer);
        }
        super.setAdapter(adapter);
        if (adapter != null) {
            adapter.registerAdapterDataObserver(observer);
        }

        checkIfEmpty();
    }

    public void setEmptyView(View emptyView) {
        mEmptyView = emptyView;
        checkIfEmpty();
    }

    public void setProgressView(ProgressBar progressBar) {
        mProgressBar = progressBar;
    }
}

File: app/src/main/java/com/chrisprime/primestationonecontrol/views/EmptyRecyclerView.kt
package com.chrisprime.primestationonecontrol.views

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.util.AttributeSet
import android.view.View

import rx.functions.Action0

class EmptyRecyclerView : RecyclerView {
    private var mEmptyView: View? = null
    private var mProgressBar: View? = null

    private var mLoading = false

    private var mEmptyAction: Action0? = null
    private var mNonEmptyAction: Action0? = null

    private val observer = object : RecyclerView.AdapterDataObserver() {
        override fun onChanged() {
            checkIfEmpty()
        }

        override fun onItemRangeInserted(positionStart: Int, itemCount: Int) {
            checkIfEmpty()
        }

        override fun onItemRangeRemoved(positionStart: Int, itemCount: Int) {
            checkIfEmpty()
        }
    }

    constructor(context: Context) : super(context) {
    }

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
    }

    constructor(context: Context, attrs: AttributeSet, defStyle: Int) : super(context, attrs, defStyle) {
    }

    fun setLoading(loading: Boolean) {
        mLoading = loading
        if (mProgressBar != null) {
            mProgressBar!!.visibility = if (loading) View.VISIBLE else View.GONE
        }
        if (loading) {
            mEmptyView!!.visibility = View.GONE
        } else {
            checkIfEmpty()
        }
    }

    fun checkIfEmpty() {
        if (adapter == null) {
            return
        }
        val emptyViewVisible = adapter.itemCount == 0
        if (mEmptyView != null) {
            mEmptyView!!.visibility = if (emptyViewVisible && !mLoading) View.VISIBLE else View.GONE
            visibility = if (emptyViewVisible || mLoading) View.GONE else View.VISIBLE
        }

        if (emptyViewVisible && mEmptyAction != null) {
            mEmptyAction!!.call()
        } else if (!emptyViewVisible && mNonEmptyAction != null) {
            mNonEmptyAction!!.call()
        }
    }


    fun setEmptyAction(emptyAction: Action0) {
        mEmptyAction = emptyAction
    }

    fun setNonEmptyAction(nonEmptyAction: Action0) {
        mNonEmptyAction = nonEmptyAction
    }

    override fun setAdapter(adapter: RecyclerView.Adapter<*>?) {
        val oldAdapter = getAdapter()
        oldAdapter?.unregisterAdapterDataObserver(observer)
        super.setAdapter(adapter)
        adapter?.registerAdapterDataObserver(observer)

        checkIfEmpty()
    }

    fun setEmptyView(emptyView: View) {
        mEmptyView = emptyView
        checkIfEmpty()
    }

    fun setProgressView(progressBar: View) {
        mProgressBar = progressBar
    }
}
